### ISSUE TITLE
Update API: support not writing history when calling external failure

### ISSUE THREAD

**Issue by longquanzheng** [temporalio/features#257]
**Is your feature request related to a problem? Please describe.**

Using “short update“ is a very important pattern of Update API feature. It acts like an RPC.  Usually, the update handler may call an external system to perform an action, if successful, then mutate the workflow states. If calling external fails, then no mutation on the workflow.

Note that this “short update“ may be called frequently with back off retry on failures. In reality, the external system could be unreliable so the failure is out of control. 

As the way it works today, Temporal will write the update failure event into history, even this event doesn’t have any value for this use case. Eventually, this failure event will blow up the history very quickly. Hence this will become a scalability issue to use with Temporal.

**Describe the solution you'd like**
Based on discussion with @mfateev , it was designed this way because throwing away the activity results means the cache would become invalid hence the workflow has to be replayed to rebuild the cache. This will put significant load on the service  as every such rejection would require kicking the workflow out of cache and replaying it from the beginning.

An better idea to implement this is to only allow certain way to calling external system -- for example, using local activity only.  In GoSDK, the update API will just change the `UpdateHandlerOptions` to become like this:
```golang
UpdateHandlerOptions struct {
    // Validator is an optional (i.e. can be left nil) func with exactly the
    // same type signature as the required update handler func but returning
    // only a single value of type error.  ....
    Validator interface{}

    // PreUpdateLocalActivity is an optional local activity to be execute after the Validator, 
    // If it succeeds, then invoke PreUpdateLocalActivityValidator if exists, or invoke the update handler
    // if it fails then return the error to Update API caller
    // Note that when this is set, the first parameter of update handler must be accepting the result of the local 
    // activity, and then the other parameters from caller
    PreUpdateLocalActivity interface{}

    // PreUpdateLocalActivityValidator is an optional func like Validator to validate the results of local activity
    // same requirement as Validator. 
    // If succeeds then invoke update handler
    // If fails then return error to Update API caller
    PreUpdateLocalActivityValidator interface{}
    
    // PreUpdateLocalActivityOptions is the options for PreUpdateLocalActivity
    PreUpdateLocalActivityOptions LocalActivityOptions
}
```

This feature will only need to implement in the SDK. And since we only update the `UpdateHandlerOptions`, it's backward compatible.  Maxim has confirmed that this should work, see the thread in the context. 

Essentially, Temporal doesn't assign any activityId(which means the state machine needs to change) or history eventId when invoking the local activity. So if it fails, nothing will be changed to the workflow states(cache). If it succeeds, the pass the results to the real update handler. 

**Describe alternatives you've considered**
Another way to workaround is to ask user to call the external system, if succeeds then passing the results when calling Update API. However, user will have to handle a lot more edge cases, especially racing conditions (as when calling external APIs, workflow is also running, and the workflow states could be changed).

**Additional context**
https://temporalio.slack.com/archives/CTDTU3J4T/p1681187414101309

**Comment by longquanzheng** [comment]
> > This feature will only need to implement in the SDK. And since we only update the UpdateHandlerOptions, it's backward compatible
> 
> I'd need server team confirmation, but this idea sounds like it needs server support as well. Since your local activity could be long running it could cause a workflow task heartbeat and the server would need to understand to throw those workflow tasks away to avoid growing your history on failure.

Yeah good call. 
but I think it’s also reasonable to limit this local activity to execute and retry within the workflow task execution.

**Comment by Quinn-With-Two-Ns** [comment]
Is the result of this `PreUpdateLocalActivity` meant to be available to the update handler?

**Comment by longquanzheng** [comment]
> Is the result of this `PreUpdateLocalActivity` meant to be available to the update handler?

Yes
> Note that when this is set, the first parameter of update handler must be accepting the result of the local  activity, and then the other parameters from caller

**Comment by bergundy** [comment]
We have a plan to allow calling "local activities" without generating history from workflows specifically for these purposes.
This mechanism would be very useful in update validators and query handlers.
It's mostly necessary for dependency injection or sandboxed workflow environments though, otherwise, you can run any code in your validation function as long as it doesn't mutate workflow state or generate history events.

**Comment by longquanzheng** [comment]
> It's mostly necessary for dependency injection or sandboxed workflow environments though, otherwise, you can run any code in your validation function as long as it doesn't mutate workflow state or generate history events.

That's not exactly the purpose. It's not just "calling local activities" but also making the results as part of the input for mutation of the workflow(in an atomic operation). Therefore running the code in validation is not going to work.

Also it's making a lot of confusion of the "readonly" requirement. And in SDKs like Java, calling external system without activities is anti-pattern of writing Temporal workflow code.

**Comment by bergundy** [comment]
I see, we had mentioned allowing the validator to mutate the input during the design of the feature.
Might be worth revisiting the decision we made where the validator can only return error or nil.

**Comment by longquanzheng** [comment]
> allowing the validator to mutate the input 

That would be nice but I don't think that's too useful unless it allows to interact with external system using [local]activities. Otherwise mutation can be done in update handler anyway.

**Comment by bergundy** [comment]
Yes, the point would be to allow interacting with external systems via an "ephemeral" local activity (the result of that local activity would not be persisted in history).

**Comment by longquanzheng** [comment]
Actually maybe I can utilize “IsReplay” api in the SDK ?

**Comment by longquanzheng** [comment]
> I see, we had mentioned allowing the validator to mutate the input during the design of the feature.
> Might be worth revisiting the decision we made where the validator can only return error or nil.

After reflecting on this, I think what is really needed is to let validator return something when passed, rather than nil.

**Comment by bergundy** [comment]
I agree this is desirable, the API is still marked experimental in Go so it's not too late to change.
Alternatively, we could let the validator add information to `UpdateInfo`.

**Comment by Quinn-With-Two-Ns** [comment]
Even if the update validator could return a result you'd still have issues with workflow task timeout, and  I don't think it is feasible in languages like Typescript with a strict sandbox for workflows.

I think we'd need to allow update validators to call local activities in some form, then we can explore allowing the validator to pass data to the handler. The second part might require a server side change
